docs/datatypes/stream/zchannel.md

```diff
@@ -0,0 +1,285 @@
+
+Let's try some examples:
+
+#### Simple Echo Channel
+
+Assume we want to read a value from the input port and then print it to the console, we can use the `ZChannel.readWith` operator to do this:
+
+```scala mdoc:compile-only
+import zio._
+import zio.stream._
+
+val producer = 
+  ZChannel.write(1)
+  
+val consumer = 
+  ZChannel.readWith(
+    (i: Int) => ZChannel.fromZIO(Console.printLine("Consumed: " + i)),
+    (_: Any) => ZChannel.unit,
+    (_: Any) => ZChannel.unit
+  )
+
+(producer >>> consumer).run
+// Output:
+// Consumed: 1
+```
+
+#### Echo Channel Forever
+
+We can also recursively compose channels to create a more complex channel. In the following example, we are going to continuously read values from the console and write them back to the console:
+
+```scala mdoc:compile-only
+import zio._
+import zio.stream.ZChannel
+
+import java.io.IOException
+
+object MainApp extends ZIOAppDefault {
+  val producer: ZChannel[Any, Any, Any, Any, IOException, String, Nothing] =
+    ZChannel
+      .fromZIO(Console.readLine("Please enter some text: "))
+      .flatMap(i => ZChannel.write(i) *> producer)
+
+  val consumer: ZChannel[Any, Any, String, Any, IOException, Nothing, Unit] =
+    ZChannel.readWith(
+      (i: String) => i match {
+        case "exit" => ZChannel.unit
+        case _ => ZChannel.fromZIO(Console.printLine("Consumed: " + i)) *> consumer
+      },
+      (_: Any) => ZChannel.unit,
+      (_: Any) => ZChannel.unit
+    )
+
+  def run = (producer >>> consumer).run
+}
+
+// Output:
+// Please enter some text: Foo
+// Consumed: Foo
+// Please enter some text: Bar
+// Consumed: Bar
+// Please enter some text: Baz
+// Consumed: Baz
+// Please enter some text: exit
+```
+
+#### Replicator Channel
+
+In this example, we are going to create a channel that replicates any input values to the output port.
+
+```scala mdoc:compile-only
+import zio._
+import zio.stream._
+
+object MainApp extends ZIOAppDefault {
+  lazy val doubler: ZChannel[Any, Any, Int, Any, Nothing, Int, Unit] =
+    ZChannel.readWith(
+      (i: Int) => ZChannel.writeAll(i, i) *> doubler,
+      (_: Any) => ZChannel.unit,
+      (_: Any) => ZChannel.unit
+    )
+  def run = (ZChannel.writeAll(1,2,3,4,5) >>> doubler).runCollect.debug
+}
+// Output:
+//   (Chunk(1,1,2,2,3,3,4,4,5,5),())
+```
+
+#### Counter Channel
+
+We can also use `Ref` to create a channel with an updatable state. For example, we can create a channel that keeps track number of all the values that it has read and finally returns it as the done value:
+
+```scala mdoc:compile-only
+import zio._
+import zio.stream._
+
+object MainApp extends ZIOAppDefault {
+  val counter =
+    ZChannel.fromZIO(Ref.make[Int](0)).flatMap { ref =>
+      lazy val inner: ZChannel[Any, Any, Int, Any, Nothing, Int, Unit] =
```

Comment:
For stateful channels, I think it could be nice (and perhaps more idiomatic and performant) to show that recursive channels can track state on the stack.
```suggestion
      def inner(currentCount: Int): ZChannel[Any, Any, Int, Any, Nothing, Int, Unit] =
      ...
        ZChannel.write(i) *> inner(currentCount + 1)
```

Reply:
Addressed. Thank you!